apps/devtools/app/scripts/components/MessageViewer/MessageTimeline/MessageTimeline.tsx

```diff
@@ -0,0 +1,143 @@
+import { Box } from '@chakra-ui/react';
+import { FC, useEffect, useMemo, useRef, useState } from 'react';
+import { Canvas, CanvasRef } from 'react-scribble';
+import { MessageTimelineMeta } from './types';
+import { TimelineDrawer } from './TimelineDrawer';
+import {
+  MessageTimelineProps,
+  RowMeta,
+  MessageSpan as MessageSpanType,
+} from './types';
+import { useTimelineColors } from './colors';
+
+const makeRowMeta = (index: number): RowMeta => ({
+  rowCursor: 0,
+  spans: [],
+  rowIndex: index,
+});
+
+export const MessageTimeline: FC<MessageTimelineProps> = ({
+  messages,
+  startedRecordingAt,
+  activeMessageId,
+  selectQuery,
+}) => {
+  /**
+   * To get this timeline to work with hovers, zooms and stuff we will needs to
+   * calculate a lot of stuff in the top level container. This would allow us to
+   * know when something is being hovered. Also the zooming of the timeline would
+   */
+  const ref = useRef<MessageTimelineMeta>({
+    startedRecordingAt,
+    currentTime: null,
+  });
+  const colors = useTimelineColors();
+  const canvas = useRef<CanvasRef<CanvasRenderingContext2D, unknown>>(null);
+  const [container, setContainer] = useState<HTMLDivElement | null>(null);
+  const [containerRect, setContainerRect] = useState<DOMRect | null>(null);
+  const { timelineSpans, rowAmount } = useMemo(() => {
```

Comment:
Would like to add some more documentation to this. This is kinda crazy / interesting sorting that probably merits some JSDocs around it.